documentation/ProjectReference-Protocol.md

```diff
@@ -119,4 +120,59 @@ These properties will then be gathered via the `GetTargetFrameworks` call.  They
+The `NearestTargetFramework` metadata will be the target framework which was selected as the best one to use for the reference (via `GetReferenceNearestTargetFrameworkTask`).  This can be used to select which set of properties were used in the target framework that was active for the reference.
+
+## SetPlatform Negotiation
+As of version 17.0, MSBuild can now dynamically figure out what platform a `ProjectReference` should build as. 
+
+### What's new?
+Modified target: `GetTargetFrameworks`
+- Extracts `$(Platform)` information and determines whether the referred project is a `.vcxproj` or `.nativeproj`.
+
+New target: `_GetProjectReferencePlatformProperties`
+- Runs after `_GetProjectReferenceTargetFrameworkProperties`
+- Calls the new `GetCompatiblePlatform` task
+- Sets or unsets `SetPlatform` metadata based on the `NearestPlatform` metadata from `GetCompatiblePlatform`.
+
+New task: `GetCompatiblePlatform`
+- Parameters: Parent's `$(Platform)`(for `.vcxproj` or `.nativeproj`) or `$(PlatformTarget)`(for managed projects), the `$(Platforms)` of the referenced project, and an optional `$(PlatformLookupTable)`
+- Using the given information, sets the `NearestPlatform` metadata as best it can or throws a warning.
+
+### How To Opt In
+First, set the properties `EnableDynamicPlatformResolution` and `DisableTransitiveProjectReferences` to `true` for **every project** in your solution. The easiest way to do this is by creating a `Directory.Build.props` file and placing it at the root of your project directory:
+
+```xml
+<Project>
+  <PropertyGroup>
+    <EnableDynamicPlatformResolution>true</EnableDynamicPlatformResolution>
+    <DisableTransitiveProjectReferences>true</DisableTransitiveProjectReferences>
+  </PropertyGroup>
+</Project>
+```
+
+ If only set in one project, the `SetPlatform` metadata will carry forward to every consecutive project reference.
+
+ Next, every referenced project is required to define a `$(Platforms)` property. `$(Platforms)` is a semicolon-delimited list of platforms that project could build as. `<Platforms>x64;x86;AnyCPU</Platforms>`, for example.
+
+ Lastly, projects that contain `ProjectReference` items may need to define a `$(PlatformLookupTable)` property. `$(PlatformLookupTable)` is a semicolon-delimited list of mappings between projects. `<PlatformLookupTable>win32=x86</PlatformLookupTable>`, for example. This means that if the current project is building for `Win32`, it should build referenced projects using `x86` as the `Platform`. This is mostly relevant for references between managed and unmanaged projects.
+
+ ### References between managed and unmanaged projects
+ Some cases of `ProjectReference`s require a `$(PlatformLookupTable)` to correctly determine what a referenced project should build as.
+| Project Reference Type | `PlatformLookupTable` Required? | Notes |
+| :--  | :-: | :-: |
+| Unmanaged -> Unmanaged | No |  |
+| Managed -> Managed | No |  |
+| Unmanaged -> Managed | **Yes** |  |
```

Comment:
Why is a table required when C++ references a C# project? C++ is always arch-specific and C# will almost always have either AnyCPU or a compatible arch-specific platform. Couldn't a default mapping of `win32=x86;x64=x64;arm=arm` apply here?

Reply:
My main confusion here was `win32` being applicable to `x86` and `AnyCPU` and not being entirely sure if either should be the default mapping from `win32`. Is win32->x86 more common?

Reply:
It goes back to my TFM analogy: where netstandard2.0 and net472 are both built by a child, a net472 parent project will pick the platform-specific one (net472) because if you offer something portable *and* a compatible specific target, the only reason the specific target would exist is because it is preferred over the portable one.
So ya, win32->x86 would be preferred over win32->anycpu by the same reasoning.
But of course for the (default) C# project that *doesn't* list x86 in its Platforms property, then the behavior would be win32->anycpu because of your "fallback to anycpu" logic. So your table for vcxproj->csproj references would just be `win32=x86`. And all other platform combination either may directly (e.g. x64=x64) or use the `*=anycpu` fallback.